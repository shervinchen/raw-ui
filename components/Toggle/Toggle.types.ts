import { ChangeEvent, InputHTMLAttributes } from "react"

interface BaseToggleProps {
  checked?: boolean
  defaultChecked?: boolean
  disabled?: boolean
  className?: string
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void
}

type NativeToggleProps = Omit<
  InputHTMLAttributes<HTMLInputElement>,
  keyof BaseToggleProps
>;

export type ToggleProps = BaseToggleProps & NativeToggleProps

export interface ToggleStyles {
  borderColor?: string;
  backgroundColor?: string;
}

export type ToggleStatus = 'unChecked' | 'checked'