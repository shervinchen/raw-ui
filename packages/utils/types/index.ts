export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends Record<string, any> ? DeepPartial<T[P]> : T[P];
};

export type Merge<T, P> = P & Omit<T, keyof P>;
