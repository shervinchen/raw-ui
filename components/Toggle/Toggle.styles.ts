import css from "styled-jsx/css";
import { RawUITheme } from "../Theme/preset/preset.type";
import { useTheme } from "../Theme/theme-context";
import { ToggleStatus, ToggleStyles } from "./Toggle.types";

const useToggleStyles = ({ status, disabled }): ToggleStyles => {
  const theme: RawUITheme = useTheme();

  const styles: {
    [key in ToggleStatus]: ToggleStyles;
  } = {
    unChecked: {
      backgroundColor: theme.palette.accents2,
      borderColor: theme.palette.accents2,
    },
    checked: {
      backgroundColor: theme.palette.black,
      borderColor: theme.palette.black,
    },
  };

  const defaultStyles = styles[status];
  const disabledStyles = {
    unChecked: {
      backgroundColor: theme.palette.accents1,
      borderColor: theme.palette.accents2,
    },
    checked: {
      backgroundColor: theme.palette.accents6,
      borderColor: theme.palette.accents6,
    },
  };

  return {
    ...defaultStyles,
    ...(disabled ? disabledStyles[status] : {}),
  };
};

export const useToggleCSS = ({ checked, disabled }) => {
  const theme: RawUITheme = useTheme();
  const { backgroundColor, borderColor } = useToggleStyles({
    status: checked ? "checked" : "unChecked",
    disabled,
  });

  return css.resolve`
    .raw-toggle {
      position: relative;
      display: inline-flex;
      cursor: ${disabled ? "not-allowed" : "pointer"};
    }
    .raw-toggle :global(.raw-toggle-input) {
      position: absolute;
      opacity: 0;
      outline: none;
      width: 0;
      height: 0;
      margin: 0;
      padding: 0;
      z-index: -1;
      font-size: 0;
      background-color: transparent;
      overflow: hidden;
      visibility: hidden;
    }
    .raw-toggle :global(.raw-toggle-inner) {
      box-sizing: border-box;
      width: 28px;
      height: 14px;
      border-radius: 14px;
      position: relative;
      transition: background 150ms cubic-bezier(0, 0, 0.2, 1);
      transition-property: background, border-color;
      background-color: ${backgroundColor};
      border: 1px solid ${borderColor};
    }
    .raw-toggle :global(.raw-toggle-inner .raw-toggle-thumb) {
      position: absolute;
      width: 12px;
      height: 12px;
      top: 50%;
      left: 0;
      box-shadow: 0 1px 2px 0 rgb(0 0 0 / 20%), 0 1px 3px 0 rgb(0 0 0 / 10%);
      transition: transform 150ms cubic-bezier(0, 0, 0.2, 1);
      border-radius: 50%;
      background-color: ${disabled
        ? theme.palette.accents2
        : theme.palette.white};
      transform: ${checked ? "translate(14px, -50%)" : "translate(0, -50%)"};
    }
  `;
};
