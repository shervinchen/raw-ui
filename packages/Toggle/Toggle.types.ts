import { ChangeEvent, ComponentPropsWithoutRef } from 'react';
import { Merge } from '../utils';

export interface BaseToggleProps {
  checked?: boolean;
  defaultChecked?: boolean;
  disabled?: boolean;
  className?: string;
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void;
}

export type ToggleProps = Merge<
  ComponentPropsWithoutRef<'input'>,
  BaseToggleProps
>;

export interface ToggleStyles {
  borderColor?: string;
  backgroundColor?: string;
}

export type ToggleStatus = 'unChecked' | 'checked';
