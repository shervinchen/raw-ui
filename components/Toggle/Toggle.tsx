import React, { FC, ChangeEvent, useEffect, useMemo } from "react";
import classNames from "classnames";
import { ToggleProps } from "./Toggle.types";
import { useControlled } from "../utils/hooks";
import { RawUITheme } from "../Theme/preset/preset.type";
import { useTheme } from "../Theme/theme-context";
import { useToggleCSS } from "./Toggle.styles";
import { VisuallyHiddenInput } from "../VisuallyHidden";

const Toggle: FC<ToggleProps> = ({
  defaultChecked = false,
  checked,
  disabled = false,
  onChange,
  className = "",
  ...restProps
}) => {
  const theme: RawUITheme = useTheme();
  const [internalValue, setInternalValue] = useControlled<boolean>({
    defaultValue: defaultChecked,
    value: checked,
  });
  const { className: resolveClassName, styles } = useToggleCSS({
    checked: internalValue,
    disabled,
  });
  const toggleClasses = classNames("raw-toggle", className, resolveClassName);
  const toggleInnerClasses = classNames("raw-toggle-inner");

  const changeHandler = (event: ChangeEvent<HTMLInputElement>) => {
    if (disabled) return;
    setInternalValue(event.target.checked);
    onChange?.(event);
  };

  return (
    <label className={toggleClasses}>
      <VisuallyHiddenInput
        className="raw-toggle-input"
        type="checkbox"
        disabled={disabled}
        checked={internalValue}
        onChange={changeHandler}
        {...restProps}
      />
      <span className={toggleInnerClasses}>
        <div className="raw-toggle-thumb" />
      </span>
      {styles}
    </label>
  );
};

export default Toggle;
