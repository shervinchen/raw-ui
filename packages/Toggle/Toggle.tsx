import React, { FC, ChangeEvent } from 'react';
import classNames from 'classnames';
import { ToggleProps } from './Toggle.types';
import { useControlled } from '../utils/hooks';
import { useToggleCSS } from './Toggle.styles';
import { VisuallyHiddenInput } from '../VisuallyHidden';

const Toggle: FC<ToggleProps> = ({
  defaultChecked = false,
  checked,
  disabled = false,
  onChange,
  className = '',
  ...restProps
}) => {
  const [internalValue, setInternalValue] = useControlled<boolean>({
    defaultValue: defaultChecked,
    value: checked,
  });
  const { className: resolveClassName, styles } = useToggleCSS({
    checked: internalValue,
    disabled,
  });
  const toggleClasses = classNames('raw-toggle', className, resolveClassName);
  const toggleInnerClasses = classNames('raw-toggle-inner');

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    setInternalValue(event.target.checked);
    onChange?.(event);
  };

  return (
    <label className={toggleClasses} data-testid="toggleLabel">
      <VisuallyHiddenInput
        className="raw-toggle-input"
        role="switch"
        type="checkbox"
        disabled={disabled}
        checked={internalValue}
        onChange={handleChange}
        {...restProps}
      />
      <span className={toggleInnerClasses}>
        <div className="raw-toggle-thumb" />
      </span>
      {styles}
    </label>
  );
};

Toggle.displayName = 'RawToggle';

export default Toggle;
