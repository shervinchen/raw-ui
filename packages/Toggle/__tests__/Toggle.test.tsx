import React, { ChangeEvent, useState } from 'react';
import { render, renderHook, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Toggle from '..';
import { ToggleProps, ToggleStatus } from '../Toggle.types';
import { useToggleStyles } from '../Toggle.styles';

const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });

describe('Toggle', () => {
  test('should match the snapshot', () => {
    const { asFragment } = render(<Toggle />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should support custom class name', () => {
    render(<Toggle className="custom-toggle" />);
    expect(screen.getByTestId('toggleLabel')).toHaveClass('custom-toggle');
  });

  test('should trigger event when clicked', async () => {
    const handleClick = jest.fn();
    render(<Toggle onClick={handleClick} />);
    const toggle = screen.getByTestId('toggleLabel');
    await user.click(toggle);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  test('should support uncontrolled value', () => {
    render(<Toggle defaultChecked />);
    const toggleInput = screen.getByRole('switch');
    expect(toggleInput).toBeChecked();
  });

  test('should support controlled value', async () => {
    const onChange = jest.fn();

    const Component = (props: ToggleProps) => {
      const [checked, setChecked] = useState(false);

      return (
        <Toggle
          checked={checked}
          onChange={(event: ChangeEvent<HTMLInputElement>) => {
            setChecked(event.target.checked);
            props.onChange?.(event);
          }}
        />
      );
    };

    render(<Component onChange={onChange} />);
    const toggle = screen.getByTestId('toggleLabel');
    const toggleInput = screen.getByRole('switch');
    expect(toggleInput).not.toBeChecked();
    await user.click(toggle);
    expect(toggleInput).toBeChecked();
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  test('should support disabled', async () => {
    render(<Toggle disabled />);
    const toggle = screen.getByTestId('toggleLabel');
    const toggleInput = screen.getByRole('switch');
    expect(toggleInput).toBeDisabled();
    await user.click(toggle);
    expect(toggleInput).not.toBeChecked();
  });

  test('should get default style when status is unknown or falsy', () => {
    const { result: result1 } = renderHook(() =>
      useToggleStyles({
        status: 'unknown' as ToggleStatus,
        disabled: false,
      })
    );

    const { result: result2 } = renderHook(() =>
      useToggleStyles({
        status: undefined as unknown as ToggleStatus,
        disabled: false,
      })
    );

    expect(result1.current.backgroundColor).toBe('#eaeaea');
    expect(result2.current.backgroundColor).toBe('#eaeaea');
  });

  test('should get default disabled style when status is unknown or falsy', () => {
    const { result: result1 } = renderHook(() =>
      useToggleStyles({
        status: 'unknown' as ToggleStatus,
        disabled: true,
      })
    );

    const { result: result2 } = renderHook(() =>
      useToggleStyles({
        status: undefined as unknown as ToggleStatus,
        disabled: true,
      })
    );

    expect(result1.current.backgroundColor).toBe('#fafafa');
    expect(result2.current.backgroundColor).toBe('#fafafa');
  });
});
