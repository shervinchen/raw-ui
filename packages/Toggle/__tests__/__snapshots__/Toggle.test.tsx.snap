// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle should match the snapshot 1`] = `
<DocumentFragment>
  <label
    class="raw-toggle jsx-2283230529"
    data-testid="toggleLabel"
  >
    <input
      class="jsx-3265204931 raw-toggle-input"
      role="switch"
      type="checkbox"
    />
    <span
      class="raw-toggle-inner"
    >
      <div
        class="raw-toggle-thumb"
      />
    </span>
  </label>
</DocumentFragment>
`;
